ext/mbstring/tests/mb_scrub.phpt

```diff
@@ -8,7 +8,15 @@ var_dump(
     "?" === mb_scrub("\x80"),
     "?" === mb_scrub("\x80", 'UTF-8')
 );
+
+$utf8str = "abc 日本語 Οὐχὶ ταὐτὰ παρίσταταί μοι γιγνώσκειν ⡍⠔⠙⠖ ⡊ ⠙⠕⠝⠰⠞";
+// Check $utf8str so it is marked as 'valid UTF-8'
+// This will enable optimized implementation of mb_scrub
+if (!mb_check_encoding($utf8str, 'UTF-8'))
+    die("Test string should be valid UTF-8");
+var_dump(mb_scrub($utf8str));
```

Comment:
Sorry for late, I have a question. This testcase `$utf8str` is not seem marked valid UTF-8. 
I ran gdb, Marked UTF-8 is works fine below case.
```
>>> r -r 'mb_scrub("a", "UTF-8");'

!5082      if (enc == &mbfl_encoding_utf8 && (GC_FLAGS(str) & IS_STR_VALID_UTF8)) {
 5083          /* A valid UTF-8 string will not be changed by mb_scrub; so just increment the refcount and return it */
 5084          RETURN_STR_COPY(str);
 5085      }

```

`mb_check_encoding` seems marks valid UTF-8 when it is not interned string.

Reply:
Thanks for pointing this out! You are right!

I don't really understand why we don't mark interned strings as valid UTF-8. I copied the test for `ZSTR_IS_INTERNED` from the PCRE extension (PCRE only marks strings as valid UTF-8 if they are not interned).

I would love to remove that test and mark interned strings as valid UTF-8 if that's what they are... but [Chesterson's fence](https://en.wiktionary.org/wiki/Chesterton%27s_fence).

For now I will add another test case like the one you showed above.

Reply:
@nikic I have just seen that you were the author of the `ZSTR_IS_INTERNED` check, in 2b9acd37f0.

Can you clarify why it is not OK to set the `IS_STR_VALID_UTF8` flag on interned strings?

Reply:
@alexdowad Interned strings are immutable. We could set the flag with an atomic rmw op. Ideally we'd just check validity of all strings during interning though.

Reply:
@nikic The concern is for ZTS interpreters running multi-threaded programs, is that right?

What if I add a `static inline` function to mark a `zend_string` as valid UTF-8, then use preprocessor directives so on ZTS builds (and for interned strings only) it uses atomic ops to set that bit, but for non-ZTS builds it just uses normal, non-atomic stores? Any concerns about that?

Regarding the idea of checking validity of *all* interned strings... I do have concerns about performance. Just benchmarked locally and the new AVX2-based UTF-8 validation takes about 13ms on my computer for a 10MB string.

The non-vectorized validation function which I just merged (derived from PCRE) takes about 200ms for a 10MB string.

Any thoughts on the performance issue?

Reply:
> The concern is for ZTS interpreters running multi-threaded programs, is that right?

I don't think this is ZTS specific. The same OPcache SHM might be shared between different processes, so any modification requires a lock.

Reply:
Right, this is not ZTS specific.

> Regarding the idea of checking validity of all interned strings... I do have concerns about performance. Just benchmarked locally and the new AVX2-based UTF-8 validation takes about 13ms on my computer for a 10MB string.

Interned strings are usually very short (at a guess, I'd expected 90% to be < 128 bytes). Of course, there's still a cost to it.

Reply:
> I don't think this is ZTS specific. The same OPcache SHM might be shared between different processes, so any modification requires a lock.

OK. Thanks for explaining that. But CMPXCHG on x86 (or equivalent on other platforms) should be good?

Reply:
> Interned strings are usually very short (at a guess, I'd expected 90% to be < 128 bytes). Of course, there's still a cost to it.

Certainly, there is a cost to it. So, is the ideal solution to implement atomic update for interned strings, or is it to validate all interned strings as UTF-8, at the time of interning?

I could (hopefully) implement either of those solutions in the next couple days, but am just trying to figure out which one to go for.

Reply:
> Certainly, there is a cost to it. So, is the ideal solution to implement atomic update for interned strings, or is it to validate all interned strings as UTF-8, at the time of interning?

I think we would need to measure for realistic cases (such as actually running some real world apps, simulating multiple concurrent clients), but that might be difficult; not sure if any of the devs has a respective enviroment available (the Windows team once had, but that is now gone; maybe @dstogov has such an environment available).

Reply:
> I think we would need to measure for realistic cases (such as actually running some real world apps, simulating multiple concurrent clients), but that might be difficult; not sure if any of the devs has a respective enviroment available (the Windows team once had, but that is now gone; maybe @dstogov has such an environment available).

If we are going to test/benchmark on "realistic cases", or even on unrealistic ones, I think it means I need to implement both solutions so performance comparisons can be done. Does that sound right? Otherwise, if neither solution has been implemented, I don't know what could actually be measured.

Reply:
> If we are going to test/benchmark on "realistic cases", or even on unrealistic ones, I think it means I need to implement both solutions so performance comparisons can be done. Does that sound right?

Yes; another drawback. :(

Reply:
> Yes; another drawback. :(

:shrug: I'm not too worried about it.